File: problems/2353_design_a_food_rating_system/2353_design_a_food_rating_system.py
import heapq

class FoodRatings:
    def __init__(self, foods, cuisines, ratings):
        self.foodToCuisine = {}
        self.foodToRating = {}
        self.cuisineToHeap = {}

        for f, c, r in zip(foods, cuisines, ratings):
            self.foodToCuisine[f] = c
            self.foodToRating[f] = r

            if c not in self.cuisineToHeap:
                self.cuisineToHeap[c] = []
            
            # push: (-rating, foodName)
            heapq.heappush(self.cuisineToHeap[c], (-r, f))

    def changeRating(self, food, newRating):
        cuisine = self.foodToCuisine[food]
        self.foodToRating[food] = newRating
        heapq.heappush(self.cuisineToHeap[cuisine], (-newRating, food))

    def highestRated(self, cuisine):
        heap = self.cuisineToHeap[cuisine]
        # Delay delete: make sure top element is latest rating
        while heap:
            rating, food = heap[0]
            if -rating == self.foodToRating[food]:
                return food
            heapq.heappop(heap)

File: problems/2353_design_a_food_rating_system/2353_design_a_food_rating_system.java
class FoodRatings {

    private Map<String, String> foodToCuisine;
    private Map<String, Integer> foodToRating;
    private Map<String, TreeSet<Food>> cuisineToFoods;

    public FoodRatings(String[] foods, String[] cuisines, int[] ratings) {
        foodToCuisine = new HashMap<>();
        foodToRating = new HashMap<>();
        cuisineToFoods = new HashMap<>();

        for (int i = 0; i < foods.length; i++) {
            String food = foods[i];
            String cuisine = cuisines[i];
            int rating = ratings[i];

            foodToCuisine.put(food, cuisine);
            foodToRating.put(food, rating);

            cuisineToFoods.putIfAbsent(cuisine, new TreeSet<>(
                    (a, b) -> a.rating == b.rating
                            ? a.name.compareTo(b.name)
                            : b.rating - a.rating));

            cuisineToFoods.get(cuisine).add(new Food(food, cuisine, rating));
        }
    }

    public void changeRating(String food, int newRating) {
        String cuisine = foodToCuisine.get(food);
        int oldRating = foodToRating.get(food);

        // Remove old value
        cuisineToFoods.get(cuisine).remove(new Food(food, cuisine, oldRating));

        // Update new value
        foodToRating.put(food, newRating);
        cuisineToFoods.get(cuisine).add(new Food(food, cuisine, newRating));
    }

    public String highestRated(String cuisine) {
        return cuisineToFoods.get(cuisine).first().name;
    }

    private class Food {
        String name;
        String cuisine;
        int rating;

        Food(String name, String cuisine, int rating) {
            this.name = name;
            this.cuisine = cuisine;
            this.rating = rating;
        }
    }
}

/**
 * Your FoodRatings object will be instantiated and called as such:
 * FoodRatings obj = new FoodRatings(foods, cuisines, ratings);
 * obj.changeRating(food,newRating);
 * String param_2 = obj.highestRated(cuisine);
 */